src/main/java/org/truffleruby/language/supercall/LookupSuperMethodNode.java

```diff
@@ -98,22 +99,26 @@ private DeclarationContext getDeclarationContext(InternalMethod currentMethod,
         final DeclarationContext context = currentMethod.getDeclarationContext();
         final DeclarationContext callerContext = currentMethod.getActiveRefinements();
 
-        if (callerContext != null) {
+        if (!DeclarationContext.isEmpty(callerContext)) {
             // super from the refined method has access to the parent's active refinements for the selfMetaClass
-            final DynamicObject[] activeRefinements = callerContext.getRefinementsFor(selfMetaClass);
+            final Map<DynamicObject, DynamicObject[]> newRefinements = new HashMap<>(context.getRefinements());
 
-            if (activeRefinements == null) {
-                return context;
-            } else {
-                final Map<DynamicObject, DynamicObject[]> newRefinements = new HashMap<>(context.getRefinements());
+            for (DynamicObject ancestor : Layouts.MODULE.getFields(selfMetaClass).ancestors()) {
```

Comment:
Honestly, this is becoming difficult in terms of readability.
In this method, we try to build the one DeclarationContext with proper refinements, and for this, we have to do:
1) one iteration by ancestors
2) one nested reverse iteration by refinements
3) Map modifications

I don't think that it dramatically affects the performance (because all the code is not executed if there are no refinements), but it's a complex code.

I suggest to remove all the method logic and instead explicitly pass `callerDeclarationContext` to the ModuleOpeartions#lookupSuperMethod.

We already have an `for` by ancestors
https://github.com/oracle/truffleruby/blob/dc80ceb7d4689069c4f0a950c243f30626c95a4d/src/main/java/org/truffleruby/core/module/ModuleOperations.java#L526-L531

and we could change the code to something like this
```java
// lexica scope refinements take precedence over parent
final DynamicObject[] refinements =
  getRefinementsFor(declarationContext, ancestor) ||  
  getRefinementsFor(callerDeclarationContext, ancestor);
```

@eregon wdyt?

Reply:
If it's just changing that line in lookupSuperMethod() it sound fine.
But I guess we need to "merge" both lexical and caller refinements?

Then I'm not sure it matters whether we merge them there or here, but probably best to deal with it in one place (in lookupSuperMethod()).

Reply:
In other words, your suggestion sounds good, could you push a commit doing that?

Reply:
> But I guess we need to "merge" both lexical and caller refinements?

Yeah, but instead of crazy (with reverse iteration) hash values merge now we are using just an array concatenation